### DISCUSSION TITLE
React hooks

### DISCUSSION THREAD

**Discussion by siddiknavadhiti** [fabricjs/fabric.js#7913] [Q&A] [answered]
Hi 
I am using fabric js in react.
In the canvas on drag of an object, particular object will be selected. On selection, how can I get object details.

The below useEffect runs on any changes on canvas, so that i can get active object. But drag and select does not consider as any changes to the canvas. I tested it by doing the console on useEffect (the console does not print if I drag and select an object).

How can I get a object, if I select by dragging ?

This is the complete code for better understanding

```
const [canvas, setCanvas] = useState('');

const initCanvas = (canvasId, width = 1125, height = 800) =>
    new fabric.Canvas(canvasId, {
      height: height,
      width: width,
      backgroundColor: '#FFFFFF',
    });
    
    useEffect(() => {
         setCanvas(initCanvas('canvas'));
     }, []);
     
    // And onclick of button I am calling the below function
    
    const onAddText = (textVal, type) => {
      let text = new fabric.IText(textVal || 'Add Text', {
      fontFamily: 'arial black',
      fontSize: type === 'h1' ? 30 : type === 'h2' ? 24 : 14,
      left: 100,
      top: 100,
      myid: "newID",
      objecttype: 'text',
    });
    // And finally adding this text to canvas 
    canvas.add(text);
  };
  
  useEffect(() => {
    console.log("Should print on any changes on the canvas")

    canvas.on('mouse:up', function(options) {
          setSelectedObj(options.target ? options.target.type : '');
    });
}, [canvas]);
```

**Comment by ShaMan123** [comment, marked as answer]
It seems you misunderstand `useEffect`. You should read about it. It will re-run only when the canvas **ref** changes, not when canvas attributes change.
You need to register events like `mouse:up`
http://fabricjs.com/events

**Reply by siddiknavadhiti** [reply]
canvas on useEffect is a state, like I have a function to create canvas which I am calling on useEffect and storing that to state.
So if a state "canvas" changes, useEffect runs

**Reply by ShaMan123** [reply]
No idea what you refer to as canvas state.
**It is NOT related to fabric anyways**.
